nntrainer/include/delegate.h

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: Apache-2.0-only
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file	delegate.h
+ * @date	7 Aug 2020
+ * @brief	This is Delegate Class for the Neural Network
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * @note This class is experimental and subject to major modifications.
+ *
+ */
+
+#ifndef __DELEGATE_H__
+#define __DELEGATE_H__
+
+namespace nntrainer {
+
+/**
+ * @class Backend
+ * @brief Backend to be used for the operations to use.
+ * @note If some operation is not supported on set backend, that operation
+ * implemented with default backend will be used.
+ *
+ */
+class Backend {
+public:
+  /**
+   * @brief Backend type enumeration
+   */
+  enum class BackendType {
+    armcl, /**< Arm Compute Library for backend */
+    blas,  /**< Libopenblas for backend */
+    base   /**< Internally implemented operations for backend */
```

Comment:
This is related with #491

Reply:
This header just stores the configuration passed by the user and validates it.
#491 implements one such backend. The structure in this header will allow allow tensors or layers to check the user set configuration and appropriately choose backend at runtime.

Reply:
I think backend type (and arguably device type... if there are two heterogenuous gpu???) should be string based as it can be added or removed from time to time.
I think we can keep this enum for now since we don't have the managing system(global context), so leaving this up to you for now :)

Reply:
custom layers are going to be really common to be done, and using string for them is really convenient.
For backends and device, string works but I will keep enum for now, and then change to string when we understand more about what various types of hardware are possible 
- different versions of devices are possible which would require a version variable (will be added in next PR)
- different types of the same group of devices are possible (npu1, npu2) - then it might be better to add enum in this case 
This all depends on how we optimize graphs with delegates and what limitations we get with layers for different devices. Till then, I will leave it as it is.

Reply:
Sounds good